### DISCUSSION TITLE
Support for Rust language

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2256] [Ideas]
I'm getting familiar with the Rust language and there is so far no full SDK when compared to ActionScript. Rust has obtained somethings near to AIR SDK, but that still don't cover everything: Amethyst and Bevy.

I heard of Flash Player projects using C++ instead of ActionScript (aka. Alchemy or CrossBridge). Considering Rust is an active language with active community and a nice package manager, did you ever consider it as a replacement for ActionScript?

If such thing were hard, I even thought of creating something similiar to AIR SDK (_[Anyone wants to team up to create a RIA SDK?](https://users.rust-lang.org/t/83441)_), but it'd be much easier if Rust could simply use AIR runtime and ADT. I've also asked on internals.rust-lang.org if the Rust team heard of AIR SDK, but it seems no one gets interest (they say it's "legacy" technology).

**UPDATE:** I'm not interested anymore in Rust.

**Comment by ajwfrost** [comment]
Hi

We'd looked into the idea of replacing the use of ActionScript with something else, but there are a host of challenges to that (and lots of objections to people who like the language!) hence are instead looking to improve ActionScript itself a little..

But you're right that it's possible to take a C++ application and turn it into a SWF file with ActionScript Byte Code sections, I saw a version of "Doom" running in the browser via Flash Player many years ago... I had wondered whether the same mechanisms can be used for other languages. As I understand it, the C++ code is compiled into LLVM, and that's then compiled into ABC (but Adobe introduced a whole load of primitive instructions for this - I'm not 100% sure whether it's actually embedded as ABC or whether you get a BinaryData object (i.e. ByteArray) with the instuctions in...)

So it ought to be possible to do the same thing with other languages, if they can be compiled into the LLVM IR. There may be some settings and configurations so some language elements/constructs perhaps wouldn't work, but as a 'raw' language it may be possible to set up.

Some interesting reading: https://dev.to/bexxmodd/llvm-infrastructure-and-rust-5g71

This could also allow other languages to be handled the same way. And it also goes the other way too - this is basically how the AOT compiler works for iPhones, the ActionScript code is passed through the `compile-abc` program that creates LLVM IR from the ActionScript Byte Code and then turns that into the appropriate native binary format for iOS.

Would be interesting to have a play with all of this...

**Comment by pfoof** [comment]
Well, it's would be the same as "Why Unity uses C# and is dependent on Microsoft, when there's Go" or "Why there's no bindings for Godot in Rust community".

As AS3 successor Haxe would be the only logical step as it is based on it and cross-compiles to C++ and compiles directly to SWF.

**Reply by hydroper** [reply]
Yeah... I think Haxe has more features than ActionScript, I was aware of it since 2016, but I'll end up not using Adobe AIR (at least for now), and maybe not even Rust (when I posted this discussion I was thinking I was going to stick to Rust, but my attempt to create something in Rust failed as the ecosystem is problematic for me).
